src/streams.h

```diff
@@ -345,18 +345,16 @@ class CDataStream
 
         // Read from the beginning of the buffer
         unsigned int nReadPosNext = nReadPos + nSize;
-        if (nReadPosNext >= vch.size())
+        if (nReadPosNext > vch.size()) {
+            throw std::ios_base::failure("CDataStream::read(): end of data");
+        }
+        memcpy(pch, &vch[nReadPos], nSize);        
```

Comment:
This is not correct. If nReadPos = 0, and nSize = 0, this will call `operator[]` on `vch`, which is illegal when it is empty.

Reply:
Note that that's a bug with the original code though, not with the refactor

Reply:
probably better to use `vch.data() + nNreadPos`?

Reply:
@sipa above there is an early return if `nSize == 0` so that doesn't happen because of the `throw`.

Reply:
Agree, there is no bug here. It would still be clearer using `vch.data() + nReadPos` though.